File: backend/src/routes/agenda.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const { getDb } = require('../db/schema');
const { autenticar, filtrarCongregacao } = require('../middleware/auth');
const { enfileirar, registrarAuditoria } = require('../services/sync');
const { getCalendarClient } = require('../services/google');

const router = express.Router();
router.use(autenticar, filtrarCongregacao);

async function criarEventoCalendar(evento) {
  try {
    const db = getDb();
    const tokensCfg = db.prepare("SELECT valor FROM configuracoes WHERE chave = 'admin_tokens'").get();
    if (!tokensCfg) return null;
    const calendar = getCalendarClient(JSON.parse(tokensCfg.valor));
    const result = await calendar.events.insert({
      calendarId: 'primary',
      requestBody: {
        summary: evento.titulo,
        description: evento.descricao || '',
        location: evento.local || '',
        start: { dateTime: evento.data_inicio, timeZone: 'America/Sao_Paulo' },
        end: { dateTime: evento.data_fim || evento.data_inicio, timeZone: 'America/Sao_Paulo' },
        reminders: { useDefault: false, overrides: [{ method: 'email', minutes: 1440 }, { method: 'popup', minutes: 60 }] },
      },
    });
    return result.data.id;
  } catch (err) {
    console.error('[Calendar] Erro ao criar evento:', err.message);
    return null;
  }
}

router.get('/', (req, res) => {
  const db = getDb();
  const filtros = ['1=1'];
  const params = [];
  if (req.congregacaoFiltro) { filtros.push('congregacao_id = ?'); params.push(req.congregacaoFiltro); }
  if (req.query.tipo) { filtros.push('tipo = ?'); params.push(req.query.tipo); }
  if (req.query.de) { filtros.push('date(data_inicio) >= ?'); params.push(req.query.de); }
  if (req.query.ate) { filtros.push('date(data_inicio) <= ?'); params.push(req.query.ate); }
  const eventos = db.prepare(`SELECT * FROM eventos WHERE ${filtros.join(' AND ')} ORDER BY data_inicio`).all(...params);
  res.json(eventos);
});

router.post('/', async (req, res) => {
  const db = getDb();
  const id = uuidv4();
  const agora = new Date().toISOString();
  const { titulo, tipo, descricao, data_inicio, data_fim, local, responsavel_email } = req.body;
  const congregacao_id = req.congregacaoFiltro || req.body.congregacao_id;
  const google_event_id = await criarEventoCalendar({ titulo, tipo, descricao, data_inicio, data_fim, local });
  const dados = { id, congregacao_id, titulo, tipo, descricao, data_inicio, data_fim, local, responsavel_email, google_event_id, criado_em: agora };
  db.prepare(`INSERT INTO eventos (id, congregacao_id, titulo, tipo, descricao, data_inicio, data_fim, local, responsavel_email, google_event_id, criado_em)
    VALUES (@id, @congregacao_id, @titulo, @tipo, @descricao, @data_inicio, @data_fim, @local, @responsavel_email, @google_event_id, @criado_em)`).run(dados);
  enfileirar('eventos', 'upsert', dados);
  registrarAuditoria(req.usuario.id, 'create', 'eventos', id, null, dados, req.ip);
  res.status(201).json(dados);
});

router.delete('/:id', (req, res) => {
  const db = getDb();
  const antes = db.prepare('SELECT * FROM eventos WHERE id = ?').get(req.params.id);
  if (!antes) return res.status(404).json({ erro: 'Não encontrado' });
  db.prepare('DELETE FROM eventos WHERE id = ?').run(req.params.id);
  enfileirar('eventos', 'delete', { id: req.params.id });
  registrarAuditoria(req.usuario.id, 'delete', 'eventos', req.params.id, antes, null, req.ip);
  res.json({ ok: true });
});

module.exports = router;

File: backend/app/routes/agenda.py
from datetime import datetime
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.deps import congregacao_filter, get_current_user
from app.models import Evento, Usuario
from app.utils import new_id, parse_datetime

router = APIRouter(prefix="/agenda", tags=["agenda"])

class EventoIn(BaseModel):
    titulo: str
    tipo: str = "culto"
    descricao: Optional[str] = None
    data_inicio: str
    data_fim: Optional[str] = None
    local: Optional[str] = None
    responsavel_email: Optional[str] = None
    congregacao_id: Optional[str] = None

class EventoOut(BaseModel):
    id: str
    congregacao_id: str
    titulo: str
    tipo: str
    descricao: Optional[str]
    data_inicio: datetime
    data_fim: Optional[datetime]
    local: Optional[str]
    responsavel_email: Optional[str]
    google_event_id: Optional[str]
    model_config = {"from_attributes": True}

@router.get("", response_model=list[EventoOut])
def listar(tipo: str = "", de: str = "", ate: str = "", db: Session = Depends(get_db),
           cu: Usuario = Depends(get_current_user), cong_filtro: Optional[str] = Depends(congregacao_filter)):
    q = db.query(Evento).filter(Evento.tenant_id == cu.tenant_id)
    if cong_filtro:
        q = q.filter(Evento.congregacao_id == cong_filtro)
    if tipo:
        q = q.filter(Evento.tipo == tipo)
    if de:
        q = q.filter(Evento.data_inicio >= parse_datetime(de))
    if ate:
        q = q.filter(Evento.data_inicio <= parse_datetime(f"{ate}T23:59:59"))
    return q.order_by(Evento.data_inicio).all()

@router.post("", response_model=EventoOut, status_code=201)
def criar(payload: EventoIn, db: Session = Depends(get_db), cu: Usuario = Depends(get_current_user),
          cong_filtro: Optional[str] = Depends(congregacao_filter)):
    cong_id = cong_filtro or payload.congregacao_id
    if not cong_id:
        raise HTTPException(status_code=400, detail="congregacao_id obrigatório")
    evento = Evento(
        id=new_id(), tenant_id=cu.tenant_id, congregacao_id=cong_id, titulo=payload.titulo,
        tipo=payload.tipo, descricao=payload.descricao, data_inicio=parse_datetime(payload.data_inicio),
        data_fim=parse_datetime(payload.data_fim), local=payload.local, responsavel_email=payload.responsavel_email,
    )
    db.add(evento); db.commit(); db.refresh(evento)
    return evento

@router.delete("/{evento_id}")
def remover(evento_id: str, db: Session = Depends(get_db), cu: Usuario = Depends(get_current_user)):
    evento = db.query(Evento).filter(Evento.id == evento_id, Evento.tenant_id == cu.tenant_id).first()
    if not evento:
        raise HTTPException(status_code=404, detail="Não encontrado")
    db.delete(evento); db.commit()
    return {"ok": True}
